Resolve to vector functions over simple functions for matching signatures

kgpai opened the pull request:
Typically we would want vector functions to take precedence in resolving over simple functions if their signatures match. The assumption here is that vector functions would be more optimized over corresponding simple functions. 

The motivation for this change comes from trying to simdize and make the 'eq' function faster. Currently 'eq' function is registered for primitive types and also supports Generic<T> which allows it to support complex types. We would like to add a simdized implementation for primitive types using a Vector function and would prefer that to be used over non optimized simple function. 

This PR also fixes a bug where we can resolve to a vector function with same name as simple function but different type signature .

mbasmanova commented on velox/expression/tests/ExprTest.cpp:
```diff
@@ -99,8 +99,64 @@ struct TestData {
+      const arg_type<Generic<T1>>& rhs) {
+    // Custom Simple Function that will always return true.
+    out = true;
+    return true;
```
Use 'void' return type.

mbasmanova reviewed (commented):
@kgpai Would you update the PR description to give an example of when we would want to register a simple and a vector function with matching signatures? 

Let's double check the behavior of the function registry APIs in the presence of "duplicate" functions: velox/functions/FunctionRegistry.h Let's make sure getFunctionSignatures doesn't return "duplicate" signatures, for example.

Let's also document the resolution order in https://facebookincubator.github.io/velox/develop/scalar-functions.html

kgpai commented:
This is a mitigation for Issue #2116 but not a complete fix.

kgpai commented:
@mbasmanova 
1. Updated the PR description .
2. This PR doesnt prevent addition of duplicates , nor does it add duplicates etc. This just changes order of resolution. We now prefer to resolve to Vector functions before simple functions. 
3. Added some documentation as per request.

kgpai commented on velox/docs/develop/scalar-functions.rst:
```diff
@@ -1309,6 +1313,19 @@ For both simple and vector functions, their names are case insensitive. Function
+
+    1. Vector Function
+    2. Simple Function with matching and specific signature
+    3. Simple Function with Generic signature
```
@laithsakka Please comment if order is wrong.

kgpai commented on velox/expression/VectorFunction.cpp:
```diff
@@ -60,11 +77,14 @@ std::shared_ptr<VectorFunction> getVectorFunction(
+      [&sanitizedName, &inputArgs, &inputTypes](
+          auto& functionMap) -> std::shared_ptr<VectorFunction> {
+        if (resolveVectorFunction(sanitizedName, inputTypes) ||
+            sanitizedName == "row_constructor") {
```
Special cased this right now, Created an issue to fix this : #2192 soon.

kgpai commented on velox/expression/VectorFunction.cpp:
```diff
@@ -37,6 +38,22 @@ std::optional<std::vector<FunctionSignaturePtr>> getVectorFunctionSignatures(
       });
 }
 
+std::shared_ptr<const Type> resolveVectorFunction(
```
Moved this to experession/VectorFunction from functions/FunctionRegistry since velox/functions depends on expressions and this creates a circular dependency.

mbasmanova commented on velox/expression/ExprCompiler.cpp:
```diff
@@ -228,6 +229,20 @@ ExprPtr getSpecialForm(
         type, std::move(compiledChildren), inputsSupportFlatNoNullsFastPath);
   }
+  if (name == kRowConstructor) {
+    static auto rowConstructorFn =
```
This code is duplicated in `if (auto concat = dynamic_cast<const core::ConcatTypedExpr*>(expr.get())) {` branch on L384. Consider extracting a helper function to avoid copy-paste.

kgpai commented:
Also addresses : #2192
